pkg/apis/seedmanagement/types_managedseed.go

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package seedmanagement
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardenletconfigv1alpha1 "github.com/gardener/gardener/pkg/gardenlet/apis/config/v1alpha1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeed represents a Shoot that is registered as Seed.
+type ManagedSeed struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	metav1.ObjectMeta
+	// Specification of the ManagedSeed.
+	Spec ManagedSeedSpec
+	// Most recently observed status of the ManagedSeed.
+	Status ManagedSeedStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeedList is a list of ManagedSeed objects.
+type ManagedSeedList struct {
+	metav1.TypeMeta
+	// Standard list object metadata.
+	metav1.ListMeta
+	// Items is the list of ManagedSeeds.
+	Items []ManagedSeed
+}
+
+// ManagedSeedSpec is the specification of a ManagedSeed.
+type ManagedSeedSpec struct {
+	// Shoot is the Shoot that will be registered as Seed.
+	Shoot *Shoot
+	// Seed describes the Seed that will be registered.
+	// Either Seed or Gardenlet must be specified.
+	Seed *SeedTemplateSpec
+	// Gardenlet specifies gardenlet deployment parameters and the GardenletConfiguration used to configure gardenlet.
+	Gardenlet *Gardenlet
+}
+
+// Shoot identifies the Shoot that will be registered as Seed.
+type Shoot struct {
+	// Name is the name of the Shoot that will be registered as Seed.
+	Name string
+	// APIServer specifies certain kube-apiserver parameters of the Shoot that will be registered as Seed.
+	APIServer *APIServer
+}
+
+// SeedTemplateSpec describes the data a Seed should have when created from a template.
+type SeedTemplateSpec struct {
```

Comment:
`SeedTemplateSpec` here follows the naming in Kubernetes, see e.g [PodTemplateSpec](https://v1-18.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#podtemplatespec-v1-core). Because of this I would propose to leave the name as it is.

Reply:
Also wanted to propose this at first, but then I noticed that the type of `appsv1.DeploymentSpec.Template` is also named `PodTemplateSpec`, so I guess, it makes sense to do it similarly here.

Reply:
Hm, still sounds strange. I guess in k/k it's not changed for backwards "compatibility" reasons, but just "template" is more appropriate as the structure also contains the object metadata. But I don't care too much about this, I'm also okay keeping it as is if you prefer this.

Reply:
Yeah, right. I also found just "Template" to be more fitting because of the metadata.

Reply:
OK, I will rename it to `SeedTemplate`. But then I think `Spec` should be left out also in other cases, e.g. `Gardenlet` and not `GardenletSpec` as proposed in another comment by @timebertt, WDYT?